servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/DefaultLoadBalancer.java

```diff
@@ -254,81 +265,71 @@ public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<R
                         DefaultLoadBalancer.this);
                 return;
             }
+            sequentialExecutor.execute(() -> sequentialOnNext(events));
+        }
 
-            boolean sendReadyEvent;
-            List<Host<ResolvedAddress, C>> nextHosts;
-            for (;;) {
-                // TODO: we have some weirdness in the event that we fail the CAS namely that we can create a host
-                //  that never gets used but is orphaned. It's fine so long as there is nothing to close but that
-                //  guarantee may not always hold in the future.
-                @SuppressWarnings("unchecked")
-                List<Host<ResolvedAddress, C>> usedHosts = usedHostsUpdater.get(DefaultLoadBalancer.this);
-                if (isClosedList(usedHosts)) {
-                    // We don't update if the load balancer is closed.
-                    return;
-                }
-                nextHosts = new ArrayList<>(usedHosts.size() + events.size());
-                sendReadyEvent = false;
-
-                // First we make a map of addresses to events so that we don't get quadratic behavior for diffing.
-                // Unfortunately we need to make this every iteration of the CAS loop since we remove entries
-                // for hosts that already exist. If this results in to many collisions and map rebuilds we should
-                // re-assess how we manage concurrency for list mutations.
-                final Map<ResolvedAddress, ServiceDiscovererEvent<ResolvedAddress>> eventMap = new HashMap<>();
-                for (ServiceDiscovererEvent<ResolvedAddress> event : events) {
-                    ServiceDiscovererEvent<ResolvedAddress> old = eventMap.put(event.address(), event);
-                    if (old != null) {
-                        LOGGER.debug("Multiple ServiceDiscoveryEvent's detected for address {}. Event: {}.",
-                                event.address(), event);
-                    }
+        private void sequentialOnNext(Collection<? extends ServiceDiscovererEvent<ResolvedAddress>> events) {
+            assert events != null && !events.isEmpty();
```

Comment:
Technically, it's possible to receive an empty collection. Not necessary to fail it

Reply:
The filtering for null or empty updates still happens on `onNext`. This was to assert that connection, but maybe it's unnecessary.